host/docs/rfnoc_image_builder.1
.TH UHD "1" "December 2024" "UHD 4" "User Commands"
.SH NAME
rfnoc_image_builder - Build UHD image using RFNoC blocks.
.SH USAGE
rfnoc_image_builder [\-h] (\fB\-y\fR YAML_CONFIG | \fB\-r\fR GRC_CONFIG) [\-F FPGA_DIR]
[\-B BUILD_DIR] [\-O BUILD_OUTPUT_DIR] [\-E BUILD_IP_DIR] [\-o IMAGE_CORE_OUTPUT]
[\-I INCLUDE_DIR] [\-b GRC_BLOCKS] [\-l LOG_LEVEL] [\-R] [\-G] [\-W] [\-S SECURE_CORE]
[\-K SECURE_KEY] [\-d DEVICE] [\-n IMAGE_CORE_NAME] [\-t TARGET] [\-g] [\-Y] [\-C] [\-s]
[\-P] [\-j JOBS] [\-c] [\-p VIVADO_PATH] [\-H] [\-D] [\-\-color {never,auto,always}]
.SS "optional arguments:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show a help message and exit
.TP
\fB\-y\fR YAML_CONFIG, \fB\-\-yaml\-config\fR YAML_CONFIG
Path to YAML configuration file (image core file). Either this option or \fB\-\-grc\-config\fR is required.
.TP
\fB\-r\fR GRC_CONFIG, \fB\-\-grc\-config\fR GRC_CONFIG
Path to .grc file to generate config from
.TP
\fB\-F\fR FPGA_DIR, \fB\-\-fpga\-dir\fR FPGA_DIR
Path to directory for the FPGA source tree. Defaults to
the FPGA source tree of the current repo.
.TP
\fB\-B\fR BUILD_DIR, \fB\-\-build\-dir\fR BUILD_DIR
Path to directory where the image core and and build artifacts will be generated. Defaults to build-<image-core-name> in the same directory as the source file.
.TP
\fB\-O\fR BUILD_OUTPUT_DIR, \fB\-\-build\-output\-dir\fR BUILD_OUTPUT_DIR
Path to directory for final FPGA build outputs. Defaults to
the FPGA's top\-level directory + /build.
.TP
\fB\-E\fR BUILD_IP_DIR, \fB\-\-build\-ip\-dir\fR BUILD_IP_DIR
Path to directory for IP build artifacts. Defaults to the
FPGA's top\-level directory + /build\-ip.
.TP
\fB\-o\fR IMAGE_CORE_OUTPUT, \fB\-\-image\-core\-output\fR IMAGE_CORE_OUTPUT
DEPRECATED! This has been replaced by \fB\-\-build\-dir\fR.
.TP
\fB\-x\fR ROUTER_HEX_OUTPUT, \fB\-\-router\-hex\-output\fR ROUTER_HEX_OUTPUT
DEPRECATED! This option will be ignored.
.TP
\fB\-I\fR INCLUDE_DIR, \fB\-\-include\-dir\fR INCLUDE_DIR
Path to directory of the RFNoC Out\-of\-Tree module
.TP
\fB\-b\fR GRC_BLOCKS, \fB\-\-grc\-blocks\fR GRC_BLOCKS
Path to directory of GRC block descriptions (needed for
\fB\-\-grc\-config\fR only)
.TP
\fB\-l\fR LOG_LEVEL, \fB\-\-log\-level\fR LOG_LEVEL
Adjust log level
.TP
\fB\-R\fR, \fB\-\-reuse\fR
Reuse existing files (do not regenerate image core).
.TP
\fB\-G\fR, \fB\-\-generate\-only\fR
Just generate files without building the FPGA
.TP
\fB\-W\fR, \fB\-\-ignore\-warnings\fR
Run build even when there are warnings in the build process.
.TP
\fB\-S\fR SECURE_CORE, \fB\-\-secure\-core\fR SECURE_CORE
Build a secure image core instead of a bitfile. This argument provides the name of the generated YAML.
.TP
\fB\-K\fR SECURE_KEY, \fB\-\-secure\-key\fR SECURE_KEY
Path to encryption key file to use for secure core.
.TP
\fB\-d\fR DEVICE, \fB\-\-device\fR DEVICE
Device to be programmed [x300, x310, e310, e320, n300, n310, n320, x410, x440]. Needs to be specified either here, or in the configuration file.
.TP
\fB\-n\fR IMAGE_CORE_NAME, \fB\-\-image\-core\-name\fR IMAGE_CORE_NAME, \fB\-\-image_core_name\fR IMAGE_CORE_NAME
Name to use for the RFNoC image core. Defaults to name of the image core YML file, without the extension.
.TP
\fB\-t\fR TARGET, \fB\-\-target\fR TARGET
Build target (e.g. X310_HG, N320_XG, ...). Needs to be specified either here, on the configuration file.
.TP
\fB\-g\fR, \fB\-\-GUI\fR
Open Vivado GUI during the FPGA building process.
.TP
\fB\-Y\fR, \fB\-\-SYNTH\fR
Stop the FPGA build process after Synthesis.
.TP
\fB\-C\fR, \fB\-\-CHECK\fR
Run elaboration only to check HDL syntax.
.TP
\fB\-s\fR, \fB\-\-save\-project\fR
Save Vivado project to disk.
.TP
\fB\-P\fR, \fB\-\-ip\-only\fR
Build only the required IPs.
.TP
\fB\-j\fR JOBS, \fB\-\-jobs\fR JOBS
Number of parallel jobs to use with make.
.TP
\fB\-c\fR, \fB\-\-clean\-all\fR
Cleans the IP before a new build.
.TP
\fB\-p\fR VIVADO_PATH, \fB\-\-vivado\-path\fR VIVADO_PATH
Path to the base install for Xilinx Vivado if not in default location (e.g., /tools/Xilinx/Vivado).
.TP
\fB\-H\fR, \fB\-\-no\-hash\fR
Do not include source YAML hash in the generated source code.
.TP
\fB\-D\fR, \fB\-\-no\-date\fR
Do not include date or time in the generated source code.
.TP
\fB\-\-color\fR {never,auto,always}
Enable colorful output. When set to 'auto' will only show color output in TTY environments (e.g., interactive shells).
.SH "DESCRIPTION"
This tool takes a YAML configuration file and generates an FPGA bitfile that
can be used with an RFNoC-capable UHD device. The YAML configuration file
describes the RFNoC blocks that are to be included in the FPGA image, along with
the connections between the blocks, and any additional configuration that is
required for the blocks.
.SH "SEE ALSO"
UHD documentation:
.B https://uhd.readthedocs.io/
.LP
Other UHD programs:
.sp
uhd_cal_tx_dc_offset(1) uhd_cal_rx_iq_balance(1) uhd_images_downloader(1)
uhd_config_info(1) uhd_find_devices(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2020 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## Next Release
* New Features:
  - Image Builder:
    - Add GRC support. This allows designing RFNoC bitfiles from GNU Radio
      Companion (GRC).
    - Improve clock connection checks and checks for duplicate connections.
    - Improve IO port compat check (e.g., check if wire widths match).
    - Allow default clocks in domain checks.
    - Simplify image core YAMLs by better usage of RADIO_NIPC parameter.
    - Improve parameter resolution.
    - Add --SYNTH and --CHECK options.
    - Add support for building an FPGA image using multiple parallel jobs and
      unique seeds with repeat_fpga_build.py and the --fpga-jobs option.
  - rfnoc-gain (OOT RFNoC example):
    - Overhaul directory structure, and rename to rfnoc-gain from
      rfnoc-example.
    - Simplify dynamic loading of OOT DLLs by using the new modules.d feature.
    - Add CE clock support to gain block. This enables the example on X4x0.
    - Add a GNU Radio subdirectory with examples of how to run the gain block
      in GNU Radio.
  - Add rfnoc_modtool. This is a command line utility to help design OOT RFNoC
    blocks.
  - RFNoC:
    - Add tune requests. This allows tuning a complete graph as known from 
      multi_usrp instead of single blocks individually.
  - multi_usrp:
    - Add Python bindings for get_user_settings_iface()
  - General UHD:
    - Add modules.d support
* Bug Fixes
  - E320: Ensure consistent sequencing when powering on/off GPSDO
  - RFNoC:
    - RFNoC DDC/DUC block (used in all Gen-3 USRPs and X410): Fix fractional
      frequency offset.
    - Fix AIS/spp calculation (e.g., for connecting FFT blocks).
  - Image Builder:
    - Fix colors
    - Fix error message for missing control SEP
    - Improve error reporting for invalid connections
    - Accept ~ and ~user on command line
    - Fix deprecated usage of yaml.load()
    - Ensure correct device tree files generation when chosing a custom build
      directory
  - rfnoc_modtool:
    - Fix generation of noc_shells
  - General UHD:
    - Fix compatibility with DPDK >= 22.11
    - Fix compiler warnings for better compatibility with C++ 17 and 20.
    - Add logic for loading uhd.dll from the correct path for Windows with Python3.8+.
    - Remove duplicate results from find
    - Release GIL when calling find from Python which improves response time for large setups.
  - MPM:
    - Allow images without RF frontend (will only be initialized if FPGA reports availability)
* Validated OS Environments
  (Versions for build and runtime dependencies can be determined from the
   docker container definitions in the UHD repository at .ci/docker/...)
  - Linux:
    - Ubuntu: bionic (18.04), focal (20.04), jammy (22.04), noble (24.04)
    - Fedora: 39, 40, 41
  - Windows: 10 21H2, 11 21H2
  - MacOS: Monterey (12.6)


## 004.007.000.000
* Highlights / Main Changes
  - Major updates to rfnoc_image_builder (a98ce26). This change adds support for
    defining transport adapters in the image builder YAML files, adding custom
    modules, and more.
  - Updated embedded OS on file system images for E3xx, N3xx, and X4xx devices
    to Yocto 4.0 (kirkstone).
* New Features
  - General
    - Allow installing only Python-based utilities (e.g., rfnoc_image_builder)
      by adding the ENABLE_PYMOD_UTILS option to CMake.
    - Added ability to query dot representation of RFNoC graph
  - Python support
    - Added DRAMReceiver class and amend rx_to_file.py example to use new class.
      This supplements the DRAMTransmitter class and simplifies using the DRAM-backed
      RFNoC replay block to when signal bandwidth exceeds available streaming bandwidth.
  - CMake
    - We now search for a system version of Pybind11 before falling back to the
      bundled version. The system version is preferred, and the bundled version
      may be removed in a future release.
  - RFNoC
    - Image builder: Major update, which adds support for defining transport
      adapters in the image builder YAML files, adding custom modules, choosing
      custom directories for build, IP, and intermediate files, building IP
      only, saving Vivado projects, and more. Image core YAML files created with
      this version are not compatible with previous versions, and vice versa!
      Also adds many more checks to avoid building bitfiles that are either
      incomplete, or valid but probably not the desired outcome.
    - Added support for defining custom register spaces in RFNoC blocks.
    - Added license checking subsystem. This allows creating RFNoC blocks that
      require a license key to operate.
    - Add device DNA feature for X310, X410, X440. This allows reading back the
      FPGA's DNA, which is a unique identifier.
  - Embedded OS for E3xx, N3xx, and X4xx devices
    - With the switch to Yocto 4.0 (kirkstone), the SD card and eMMC images no 
      longer pre-populate the second system partition. Refer to the device 
      specific "Updating Filesystems" section in the user manual for how to populate it.
      This reduces the size of compressed .sdimg filesystem images.
  - X440
    - Added new guided log info mode to provide more information about the Nyquist
      zone in which the chosen frequency is located. These additional info logs
      are enabled by default. For disabling the guided mode, add the entry
      `guided_mode=false` to the UHD config file in the "Global" section.
  - Examples
    - rfnoc-example now also generates Python bindings for OOT block control functions.
* Dependencies changes
  - Increase minimum required version
    - Python to 3.7.0
    - Pybind11 to 2.7.0
* Bug Fixes
  - General UHD
    - Fix issue with undefined HOME or XDG_CONFIG_HOME
    - Fix compilation error with boost 1.85.0 (closed issue 740)
    - Fixed cmake find_library for UHD to fail if the library is not found
    - Fixed setters for ALL_CHANS in multi_usrp
    - Demoted safe call logging to warning

host/utils/rfnoc_image_builder.py
#!/usr/bin/env python3
"""RFNoC Image Builder: Create RFNoC FPGA bitfiles from YAML input.

SPDX-License-Identifier: GPL-3.0-or-later
"""

import argparse
import hashlib
import logging
import os
import re
import sys

from uhd.rfnoc_utils import image_builder, log as rfnoc_log


def setup_parser():
    """Create argument parser."""
    parser = argparse.ArgumentParser(
        description="Build UHD image using RFNoC blocks",
    )

    config_group = parser.add_mutually_exclusive_group(required=True)
    config_group.add_argument("-y", "--yaml-config", help="Path to yml configuration file")
    config_group.add_argument("-r", "--grc-config", help="Path to grc file to generate config from")
    parser.add_argument(
        "-C",
        "--base-dir",
        help="Path to the base directory. Defaults to the current directory.",
        required=False,
        default=os.getcwd(),
    )
    parser.add_argument(
        "-F",
        "--fpga-dir",
        help="Path to directory for the FPGA source tree. "
        "Defaults to the FPGA source tree of the current repo.",
        required=False,
        default=None,
    )
    parser.add_argument(
        "-B",
        "--build-dir",
        help="Path to directory where the image core and and build artifacts will be generated. "
        "Defaults to \"build-<image-core-name>\" in the base directory.",
        required=False,
        default=None,
    )
    parser.add_argument(
        "-O",
        "--build-output-dir",
        help="Path to directory for final FPGA build outputs. "
        "Defaults to \"build\" in the base directory.",
        required=False,
        default=None,
    )
    parser.add_argument(
        "-E",
        "--build-ip-dir",
        help="Path to directory for IP build artifacts. "
        "Defaults to \"build-ip\" in the base directory.",
        required=False,
        default=None,
    )
    parser.add_argument(
        "-o", "--image-core-output", help="DEPRECATED! This has been replaced by --build-dir. "
    )
    parser.add_argument(
        "-x", "--router-hex-output", help="DEPRECATED! This option will be ignored. ", default=None
    )
    parser.add_argument(
        "-I",
        "--include-dir",
        help="Path to directory of the RFNoC Out-of-Tree module",
        action="append",
        default=[],
    )
    parser.add_argument(
        "-b",
        "--grc-blocks",
        help="Path to directory of GRC block descriptions (needed for --grc-config only)",
        default=None,
    )
    parser.add_argument("-l", "--log-level", help="Adjust log level", default="info")
    parser.add_argument(
        "-R",
        "--reuse",
        help="Reuse existing files (do not regenerate image core).",
        action="store_true",
    )
    parser.add_argument(
        "-G",
        "--generate-only",
        help="Just generate files without building the FPGA",
        action="store_true",
    )
    parser.add_argument(
        "-W",
        "--ignore-warnings",
        help="Run build even when there are warnings in the build process",
        action="store_true",
    )
    parser.add_argument(
        "-S",
        "--secure-core",
        help="Build a secure image core instead of a bitfile. "
        "This argument provides the name of the generated YAML.",
    )
    parser.add_argument(
        "-K",
        "--secure-key",
        help="Path to encryption key file to use for secure core.",
        default="",
    )
    parser.add_argument(
        "-d",
        "--device",
        help="Device to be programmed [x300, x310, e310, e320, n300, n310, n320, x410, x440]. "
        "Needs to be specified either here, or in the configuration file.",
        default=None,
    )
    parser.add_argument(
        "-n",
        "--image-core-name",
        "--image_core_name",
        help="Name to use for the RFNoC image core. "
        "Defaults to name of the image core YML file, without the extension.",
        default=None,
    )
    parser.add_argument(
        "-t",
        "--target",
        help="Build target (e.g. X310_HG, N320_XG, ...). Needs to be specified "
        "either here, on the configuration file.",
        default=None,
    )
    parser.add_argument(
        "-g", "--GUI", help="Open Vivado GUI during the FPGA building process", action="store_true"
    )
    parser.add_argument(
        "-Y",
        "--SYNTH",
        help="Stop the FPGA build process after Synthesis",
        action="store_true",
    )
    parser.add_argument(
        "--CHECK",
        help="Run elaboration only to check HDL syntax",
        action="store_true",
    )
    parser.add_argument(
        "-s", "--save-project", help="Save Vivado project to disk", action="store_true"
    )
    parser.add_argument("-P", "--ip-only", help="Build only the required IPs", action="store_true")
    parser.add_argument(
        "-j",
        "--jobs",
        help="Number of parallel jobs to use with make",
        required=False,
        default=None,
    )
    parser.add_argument(
        "-c", "--clean-all", help="Cleans the IP before a new build", action="store_true"
    )
    parser.add_argument(
        "-p",
        "--vivado-path",
        help="Path to the base install for Xilinx Vivado if not in default "
        "location (e.g., /tools/Xilinx/Vivado).",
        default=None,
    )
    parser.add_argument(
        "-H",
        "--no-hash",
        help="Do not include source YAML hash in the generated source code.",
        action="store_true",
        default=False,
    )
    parser.add_argument(
        "-D",
        "--no-date",
        help="Do not include date or time in the generated source code.",
        action="store_true",
        default=False,
    )
    parser.add_argument(
        "--color",
        choices=("never", "auto", "always"),
        default="auto",
        help="Enable colorful output. When set to 'auto' will only show color "
        "output in TTY environments (e.g., interactive shells)",
    )

    return parser


def resolve_path(path, local):
    """Replace path by local if path is enclosed with "@" (placeholder markers).

    :param path: the path to check
    :param local: new path content if path is placeholder
    :return: path if path is not a placeholder else local
    """
    return re.sub("^@.*@$", local, path)


def get_fpga_path(args):
    """Return FPGA path.

    This is the fpga_dir from the arguments. If fpga_dir
    does not exist, it is changed to the FPGA source path of the current repo.
    If current directory is not in a repo, an error is generated.
    :param args: arguments passed to the script
    :return: FPGA root path
    """
    # If a valid path is given, use that
    if args.fpga_dir:
        fpga_path = os.path.normpath(os.path.abspath(os.path.expanduser(args.fpga_dir)))
        if os.path.isdir(fpga_path):
            logging.info("Using FPGA directory %s", fpga_path)
            return fpga_path
        else:
            logging.error("Bad fpga-dir argument. %s is not a valid directory.", args.fpga_dir)
            sys.exit(1)
    # Try to find an fpga directory at the base of the repo
    repo_base = os.getcwd()
    while True:
        repo_base = os.path.split(repo_base)[0]
        top_dir = os.path.split(repo_base)[1]
        if top_dir == "uhd" or top_dir == "uhddev":
            fpga_path = os.path.join(repo_base, "fpga")
            break
        if top_dir == "":
            fpga_path = None
            break
    # If it's valid FPGA base path, use that
    if fpga_path and os.path.isdir(os.path.join(fpga_path, "usrp3", "top")):
        logging.info("Using FPGA directory %s", fpga_path)
        return fpga_path
    # No valid path found
    logging.error("FPGA path not found. Specify with --fpga-dir argument.")
    sys.exit(1)


def get_config_path():
    """Return path that contains configurations files.

    This is the main UHD installation path for package data (e.g., /usr/share/uhd).

    Will return the directory path where the YAML configuration files are stored
    (yml descriptions for block, IO signatures and device bsp, not the image
    core files).

    :return: Configuration path
    """
    return os.path.normpath(
        resolve_path(
            "@CONFIG_PATH@", os.path.join(os.path.dirname(__file__), "..", "include", "uhd")
        )
    )


def main():
    """Run image_builder.build_image.

    :return: exit code
    """
    args = setup_parser().parse_args()
    rfnoc_log.init_logging(args.color, args.log_level)

    source_arg = args.yaml_config if args.yaml_config else args.grc_config
    source = os.path.normpath(os.path.abspath(os.path.expanduser(source_arg)))
    source_type = "yaml" if args.yaml_config else "grc"
    if not os.path.isfile(source):
        logging.error("Source file %s does not exist", source)
        return 1

    source_hash = hashlib.sha256()
    with open(source, "rb") as source_file:
        source_hash.update(source_file.read())

    dump_config = args.grc_config and not args.yaml_config

    # Do some more sanitization of command line arguments
    if args.image_core_output:
        logging.warning(
            "Using deprecated command line option --image-core-output (-o)! "
            "Use --build-dir instead."
            "This option will be removed in future versions of UHD."
        )
        if args.build_dir:
            logging.warning(
                "Both --build-dir and --image-core-output are used. Ignoring the latter."
            )
        else:
            args.build_dir = args.image_core_output

    return image_builder.build_image(
        # Image core source file info
        source=source,
        source_type=source_type,
        source_hash=source_hash.hexdigest(),
        no_date=args.no_date,
        no_hash=args.no_hash,
        # Unsanitized device ID string from cmd line (e.g., x410, n320, ...)
        device=args.device,
        # Build target info (unsanitized, from command line)
        image_core_name=args.image_core_name,
        target=args.target,
        # Provide all the paths
        base_dir = args.base_dir,
        build_dir=args.build_dir,
        build_output_dir=args.build_output_dir,
        build_ip_dir=args.build_ip_dir,
        repo_fpga_path=get_fpga_path(args),
        config_path=get_config_path(),
        yaml_path=args.yaml_config if args.yaml_config else args.grc_config,
        include_paths=args.include_dir,
        vivado_path=args.vivado_path,
        grc_blocks=args.grc_blocks,
        # Various build config parameters
        reuse=args.reuse,
        generate_only=args.generate_only,
        continue_on_warnings=args.ignore_warnings,
        clean_all=args.clean_all,
        GUI=args.GUI,
        synthesize_only=args.SYNTH,
        check_hdl=args.CHECK,
        save_project=args.save_project,
        ip_only=args.ip_only,
        num_jobs=args.jobs,
        secure_core=args.secure_core,
        secure_key_file=args.secure_key,
        dump_config=dump_config,
    )


if __name__ == "__main__":
    ret_val = main()
    # In some cases, we can get return codes out of the standard 0-254 range.
    # Coerce it the standard value of 255 to indicate it was out of range.
    if ret_val > 255:
        ret_val = 255
    sys.exit(ret_val)
